### DISCUSSION TITLE
Is the http URL scheme not supported in #include <url>?

### DISCUSSION THREAD

**Discussion by falbrechtskirchinger** [compiler-explorer/compiler-explorer#3548] [Q&A] [answered]
I'm trying to include a file hosted on my local machine. Previously, I've just pushed to GitHub and used "raw.githubusercontent.com", now I'm looking to eliminate that step. First, IP addresses (v4 or v6) seem to not be supported, right? Port numbers apparently are. I did create an AAAA record for my IPv6 address and even hosted the file on port 80 when that still wouldn't work. I'm trying to avoid setting up a certificate for 5 minutes of debugging...

I've obviously made sure the file is fetchable outside my network and I won't share it because it is my home IP address.

(Also, I looked at the code … could someone please point me to where `#include <url>` is handled? I was unable to find it. Thanks.)

**Comment by AbrilRBS** [comment, marked as answer]
The expansion is handled client-side @ https://github.com/compiler-explorer/compiler-explorer/blob/main/static/compiler-service.js#L269
Note how we specify that the protocol should be https. We could maybe relax this restriction, but even if nothing else restricts it (Such as our Content-Security-Policy), you'd still need to at the very least set a CORS header so that you can fetch it.

Still, have you tried the Tree feature? While it does not allow for changes of the file on your side to be atuomatically picked up, you can add more than 1 file to the compilation that way, and using the "Save/Load" button on the second editor would let you upload the file directly from your fs.

**Reply by falbrechtskirchinger** [reply]
> The expansion is handled client-side @ https://github.com/compiler-explorer/compiler-explorer/blob/main/static/compiler-service.js#L269 Note how we specify that the protocol should be https. We could maybe relax this restriction, but even if nothing else restricts it (Such as our Content-Security-Policy), you'd still need to at the very least set a CORS header so that you can fetch it.

Well that explains a lot. Thanks! I did see a note about cross-site scripting and didn't understand how that would apply. Now I do.

> Still, have you tried the Tree feature? While it does not allow for changes of the file on your side to be atuomatically picked up, you can add more than 1 file to the compilation that way, and using the "Save/Load" button on the second editor would let you upload the file directly from your fs.

I'm aware but I'm really looking for a solution that picks up changes immediately. Guess the only option is to setup a certificate anyway.

**Comment by apmorton** [comment]
Give ngrok a try:
https://ngrok.com/

It gives you an ephemeral domain with https for free and requires no port
forwarding.

Just run `ngrok http 80` on your machine (where your server is listening on
localhost:80) and it will spit out a public https URL for you.

**Reply by falbrechtskirchinger** [reply]
I use ngrok for tunneling SSH connections. I had no idea it could do that. :man_facepalming:

But it turns out creating a certificate is as easy as `certbot certonly --standalone` these days, so it wasn't a big problem after all.

Thanks anyway!
